### ISSUE TITLE
improve inbound filters

### ISSUE THREAD

**Issue by Niryo** [getsentry/sentry#80731]
### Problem Statement

In mobile, we have many requests that are coming from outdated clients, and it makes it hard to understand which errors are not relevant anymore (we keep getting errors that were already fixed in newer clients).
Our current way of overcoming this issue is to use inbound filters to filter out versions from old clients. The issue with that approach is that inbound filters are matched as strings, and we can't add filters like this for example: **> 2.100155.0**, in order to filter out all releases that are less than version 2.100155.0. 
So in order to achieve something like that, we have some complex inbound filters that we update on every new release, and it looks like this: 
com.app@2.?.*+*
com.app@2.??.*+*
com.app@2.???.*+*
com.app@2.????.*+*
com.app@2.?????.*+*
com.app@[0-1].??????.?+*
com.app@2.0?????.?+*
com.app@2.1000??.?+*
com.app@2.1001[0-4]?.?+*
com.app@2.10015[0-4].?+*

Not pleasant at all as you can see.

### Solution Brainstorm

One suggested solution would be to support richer filter syntax, like suggested above.
Another option is to add whiteList filters instead of blacklist, so we could just say "filter out everything which isn't coming from this release and that release.

### Product Area

Ingestion and Filtering

**Comment by jangjodi** [comment]
Thank you for the suggestion! I have added this to our backlog, though I am unsure when this can be prioritized.

**Comment by Dav1dde** [comment]
Would an inverted glob pattern already solve this for you? E.g. a syntax like `!com.app@1.2.3` filtering everything except the version `com.app@1.2.3`?

**Comment by Niryo** [comment]
It depends, what would happen if we'll do
```
!1.1.0
!1.2.0
```

Will it allow both 1.1.0 and 1.2.0 to pass? 
If so then yes, it might help improve our current crazy inbound filter descriptor. It will still be a bit tricky because we'll need to update the filters as soon as a new version is out, so something like `<1.1.0` would be better, but it's something to start with.

**Comment by Dav1dde** [comment]
> Will it allow both 1.1.0 and 1.2.0 to pass?

No, but you could express it as something like `!{1.1.0,1.2.0}`.

I created a [Relay issue](https://github.com/getsentry/relay/issues/4286) to track this separately, since that isn't necessarily related to your feature request.

> It will still be a bit tricky because we'll need to update the filters as soon as a new version is out, so something like <1.1.0 would be better, but it's something to start with.

Yeah definitely agree that there should be some more powerful matching.

**Comment by Niryo** [comment]
also, I want to raise another concern that I have regarding inbound filters- we are soon going to support code-push in our app, which means that for the same native app, we'll have many different virtual versions. I'll explain:
we have multiple js bundles, and a single native app. we upload the native app to the stores, and when a user launches the native app, we downloadsthe js bundles dynamically. so there could be two users that have version `1.0.0` installed on their device, but one user will have `bundleA` with a version of `2.2.2` and another user will have `bundleA` with a greater version, say `3.3.3`. but sentry allows us to filter only based on the native version, which means we'll still get errors from `bundleA v2.2.2`, even though those error are fixed in `v3.3.3` and are not relevant anymore.

I am not sure how sentry can handle this use case...maybe you can add a hook for inbound filter, and call our endpoint with the event and we will return true/false if the event should be filtered or not? we already adding to every event a custom property that contains all the versions of all the js bundles that where downloaded, so on our side we could filter out the event based on this prop using our custom logic.
I guess that the issue here is that it might be expensive to call our endpoint for every event, but I can't think of a way to add custom filtering logic without you implementing some more complex filtering language, and give us access to check custom props, and not only the release prop.

**Comment by Dav1dde** [comment]
Unfortunately I cannot give you any timeline for this since I work on different parts of Sentry, but hopefully the the hard parts are already done.

**Comment by alfredocambera-changeorg** [comment]
This would make filtering versions way simpler than the current alternative of using glob expressions. I envision this working like terraform's https://developer.hashicorp.com/terraform/language/expressions/version-constraints

**Comment by alfredocambera-changeorg** [comment]
any progress on this feature?
